src/lua/uri.c

```diff
@@ -0,0 +1,313 @@
+ */
+#include "lua/uri.h"
+
+#include "lua/utils.h"
+#include "uri/uri.h"
+#include "diag.h"
+
+/**
+ * Add values to URI query parameter @a param from lua table, which located
+ * at the top of the lua stack. All not numeric keys are  treaded as errors.
+ * Allowed types for URI query parameter values are LUA_TSTRING and LUA_TNUMBER.
+ */
+static int
+uri_add_param_from_lua_table(struct uri_query_param *param, struct lua_State *L)
+{
+	int rc = 0;
+	lua_pushnil(L);
+	while (lua_next(L, lua_gettop(L) - 1) != 0 && rc == 0) {
+		if (lua_type(L, -2) == LUA_TNUMBER) {
+			int type = lua_type(L, -1);
+			if (type == LUA_TSTRING || type == LUA_TNUMBER) {
+				const char *val = lua_tostring(L, -1);
+				uri_query_param_add_value(param, val);
+			} else {
+				rc = -1;
+			}
+		} else {
+			rc = -1;
+		}
+		lua_pop(L, 1);
+	}
+	return rc;
+}
+
+/**
+ * Add URI query parameter with @a param_name to @a uri from the value
+ * at the top of the lua stack. Allowed types for URI query parameter
+ * values are LUA_TSTRING, LUA_TNUMBER and LUA_TTABLE.
+ */
+static int
+uri_add_param_from_lua(struct uri *uri, struct lua_State *L, const char *name)
+{
+	int rc = 0;
+	struct uri_query_param *param =
+		uri_find_query_param(uri, name);
+	if (param == NULL)
+		param = uri_add_query_param(uri, name);
+	int type = lua_type(L, -1);
+	switch (type) {
+	case LUA_TSTRING:
+	case LUA_TNUMBER:
+		uri_query_param_add_value(param, lua_tostring(L, -1));
+		break;
+	case LUA_TTABLE:
+		rc = uri_add_param_from_lua_table(param, L);
+		break;
+	default:
+		rc = -1;
+	}
+	return rc;
+}
+
+/**
+ * Add URI query parameter with @a name to all URIs in @a uri_set from the
+ * value at the top of the lua stack.
+ */
+static int
+uri_set_add_param_from_lua(struct uri_set *uri_set, struct lua_State *L,
+			   const char *name, bool if_not_exist)
+{
+	int rc = 0;
+	for (int i = 0; i < uri_set->uri_count && rc == 0; i++) {
+		struct uri *uri = &uri_set->uris[i];
+		if (uri_query_param_count(uri, name) != 0 && if_not_exist)
+			continue;
+		rc = uri_add_param_from_lua(uri, L, name);
+	}
+	return rc;
+}
+
+/**
+ * Add or overwrite (depends on @a param_is_overwritten) URI query parameters
+ * for all URIs in @a uri_set. Place the table appropriate to the @a table_name
+ * key on the top of the lua stack and use it to get the list of URI query
+ * parameters and there new values.
+ */
+static int
+uri_set_add_params_from_lua(struct uri_set *uri_set, struct lua_State *L,
+			    const char *table_name, bool param_is_overwritten)
+{
+	lua_pushstring(L, table_name);
+	lua_rawget(L, -2);
+	if (lua_type(L, -1) == LUA_TNIL) {
+		lua_pop(L, 1);
+		return 0;
+	} else if (lua_type(L, -1) != LUA_TTABLE) {
+		lua_pop(L, 1);
+		return -1;
+	}
+	int rc = 0;
+	lua_pushnil(L);
+	while (lua_next(L, lua_gettop(L) - 1) != 0 && rc == 0) {
+		if (lua_type(L, -2) == LUA_TSTRING) {
+			const char *param_name = lua_tostring(L, -2);
+			bool if_not_exist = !param_is_overwritten;
```

Comment:
Please use `if_not_exists` argument instead of `param_is_overwritten`. Using two opposite flags for the same thing is confusing.

Reply:
fixed